File: Sem_5/Mobile/non-native/BooklogFlutter/lib/ui/screens/edit_book_dialog.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../../data/book_review_entry.dart';
import '../components/form_components.dart';

class EditBookDialog extends StatefulWidget {
  final BookReviewEntry book;
  final VoidCallback onDismiss;
  final Function(BookReviewEntry) onConfirm;

  const EditBookDialog({
    super.key,
    required this.book,
    required this.onDismiss,
    required this.onConfirm,
  });

  @override
  State<EditBookDialog> createState() => _EditBookDialogState();
}

class _EditBookDialogState extends State<EditBookDialog> {
  late String _coverImageURL;
  late String _bookTitle;
  late String _genre;
  late String _authorName;
  late String _readingStatus;
  late DateTime _startDate;
  late String _starRating;
  late String _reviewDescription;

  String? _bookTitleError;
  String? _authorNameError;
  String? _starRatingError;

  @override
  void initState() {
    super.initState();
    _coverImageURL = widget.book.coverImageURL;
    _bookTitle = widget.book.bookTitle;
    _genre = widget.book.genre;
    _authorName = widget.book.authorName;
    _readingStatus = widget.book.readingStatus;
    _startDate = widget.book.startDate;
    _starRating = widget.book.starRating > 0
        ? widget.book.starRating.toString()
        : "";
    _reviewDescription = widget.book.reviewDescription;
  }

  Future<void> _selectDate(BuildContext context) async {
    final DateTime? picked = await showDatePicker(
      context: context,
      initialDate: _startDate,
      firstDate: DateTime(2000),
      lastDate: DateTime(2100),
    );
    if (picked != null && picked != _startDate) {
      setState(() {
        _startDate = picked;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    final dateFormat = DateFormat('MMM dd, yyyy');

    return Dialog(
      child: Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          title: const Text(
            "Edit Book",
            style: TextStyle(
              color: Colors.black,
              fontSize: 18,
              fontWeight: FontWeight.normal,
            ),
          ),
          backgroundColor: Colors.white,
          elevation: 0,
          leading: IconButton(
            icon: const Icon(Icons.arrow_back, color: Colors.black),
            onPressed: widget.onDismiss,
          ),
          actions: [
            TextButton(
              onPressed: widget.onDismiss,
              child: const Text(
                "Cancel",
                style: TextStyle(
                  color: Color(0xFFE53935),
                  fontSize: 16,
                ),
              ),
            ),
          ],
        ),
        body: SingleChildScrollView(
          padding: const EdgeInsets.all(16),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              // Cover Image URL
              CustomFormField(
                label: "Cover Image (URL)",
                value: _coverImageURL,
                onValueChange: (value) {
                  setState(() {
                    _coverImageURL = value;
                  });
                },
                placeholder: "Cover Image URL",
              ),
              const SizedBox(height: 20),
              // Book Title
              CustomFormField(
                label: "Book Title",
                value: _bookTitle,
                onValueChange: (value) {
                  setState(() {
                    _bookTitle = value;
                    _bookTitleError = null;
                  });
                },
                placeholder: "Book Title",
                isError: _bookTitleError != null,
                errorMessage: _bookTitleError,
              ),
              const SizedBox(height: 20),
              // Genre
              CustomFormField(
                label: "Genre",
                value: _genre,
                onValueChange: (value) {
                  setState(() {
                    _genre = value;
                  });
                },
                placeholder: "Book genre",
              ),
              const SizedBox(height: 20),
              // Author Name
              CustomFormField(
                label: "Author Name",
                value: _authorName,
                onValueChange: (value) {
                  setState(() {
                    _authorName = value;
                    _authorNameError = null;
                  });
                },
                placeholder: "Book Author",
                isError: _authorNameError != null,
                errorMessage: _authorNameError,
              ),
              const SizedBox(height: 20),
              // Status
              const Text(
                "Status",
                style: TextStyle(
                  fontSize: 14,
                  color: Colors.black,
                ),
              ),
              const SizedBox(height: 8),
              StatusSelector(
                selectedStatus: _readingStatus,
                onStatusSelected: (status) {
                  setState(() {
                    _readingStatus = status;
                  });
                },
              ),
              const SizedBox(height: 20),
              // Start Date
              ActionFormField(
                label: "Start date",
                value: dateFormat.format(_startDate),
                placeholder: "Start date of reading",
                onTap: () => _selectDate(context),
                trailingIcon: Icons.calendar_today_outlined,
              ),
              const SizedBox(height: 20),
              // Rating
              CustomFormField(
                label: "Rating (0 - 5 stars)",
                value: _starRating,
                onValueChange: (value) {
                  setState(() {
                    _starRating = value;
                    _starRatingError = null;
                  });
                },
                placeholder: "Rating",
                isError: _starRatingError != null,
                errorMessage: _starRatingError,
              ),
              const SizedBox(height: 20),
              // Rating Description
              CustomFormField(
                label: "Rating Description",
                value: _reviewDescription,
                onValueChange: (value) {
                  setState(() {
                    _reviewDescription = value;
                  });
                },
                placeholder: "Description",
                minLines: 3,
              ),
              const SizedBox(height: 16),
              // Update Button
              SizedBox(
                width: double.infinity,
                child: ElevatedButton(
                  onPressed: () {
                    // Validation
                    bool isValid = true;
                    setState(() {
                      _bookTitleError = null;
                      _authorNameError = null;
                      _starRatingError = null;
                    });

                    if (_bookTitle.trim().isEmpty) {
                      setState(() {
                        _bookTitleError = "Book title is required";
                      });
                      isValid = false;
                    }

                    if (_authorName.trim().isEmpty) {
                      setState(() {
                        _authorNameError = "Author name is required";
                      });
                      isValid = false;
                    }

                    if (_starRating.isNotEmpty) {
                      final rating = double.tryParse(_starRating);
                      if (rating == null || rating < 0 || rating > 5) {
                        setState(() {
                          _starRatingError = "Rating must be between 0 and 5";
                        });
                        isValid = false;
                      }
                    }

                    if (isValid) {
                      final updatedBook = widget.book.copyWith(
                        bookTitle: _bookTitle.trim(),
                        authorName: _authorName.trim(),
                        genre: _genre.trim(),
                        readingStatus: _readingStatus,
                        starRating: _starRating.isNotEmpty
                            ? double.tryParse(_starRating) ?? 0.0
                            : 0.0,
                        reviewDescription: _reviewDescription.trim(),
                        coverImageURL: _coverImageURL.trim(),
                        startDate: _startDate,
                      );
                      widget.onConfirm(updatedBook);
                    }
                  },
                  style: ElevatedButton.styleFrom(
                    backgroundColor: const Color(0xFF2196F3),
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(8),
                    ),
                    padding: const EdgeInsets.symmetric(vertical: 12),
                  ),
                  child: const Text(
                    "Update Review",
                    style: TextStyle(fontSize: 16),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: Sem_5/Mobile/native/Booklog/app/src/main/java/com/example/booklog/ui/screens/EditBookDialog.kt
package com.example.booklog.ui.screens

import android.app.DatePickerDialog
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import com.example.booklog.data.BookReviewEntry
import com.example.booklog.ui.components.FormField
import com.example.booklog.ui.components.StatusSelector
import java.text.SimpleDateFormat
import java.util.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EditBookDialog(
    book: BookReviewEntry,
    onDismiss: () -> Unit,
    onConfirm: (BookReviewEntry) -> Unit
) {
    val context = LocalContext.current
    var coverImageURL by remember { mutableStateOf(book.coverImageURL) }
    var bookTitle by remember { mutableStateOf(book.bookTitle) }
    var genre by remember { mutableStateOf(book.genre) }
    var authorName by remember { mutableStateOf(book.authorName) }
    var readingStatus by remember { mutableStateOf(book.readingStatus) }
    var starRating by remember { mutableStateOf(if (book.starRating > 0) book.starRating.toString() else "") }
    var reviewDescription by remember { mutableStateOf(book.reviewDescription) }
    var startDate by remember { mutableStateOf(book.startDate) }
    val calendar = remember { Calendar.getInstance() }
    val dateFormat = remember { SimpleDateFormat("MMM dd, yyyy", Locale.getDefault()) }

    var bookTitleError by remember { mutableStateOf<String?>(null) }
    var authorNameError by remember { mutableStateOf<String?>(null) }
    var starRatingError by remember { mutableStateOf<String?>(null) }
    
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false)
    ) {
        Surface(
            modifier = Modifier.fillMaxSize(),
            color = Color.White
        ) {
            Column(
                modifier = Modifier.fillMaxSize()
            ) {
                // Header
                TopAppBar(
                    title = { 
                        Text(
                            "Edit Book",
                            color = Color.Black,
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Normal
                        )
                    },
                    navigationIcon = {
                        IconButton(onClick = onDismiss) {
                            Icon(
                                Icons.Default.ArrowBack,
                                contentDescription = "Back",
                                tint = Color.Black
                            )
                        }
                    },
                    actions = {
                        TextButton(onClick = onDismiss) {
                            Text(
                                "Cancel",
                                color = Color(0xFFE53935),
                                fontSize = 16.sp
                            )
                        }
                    },
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = Color.White
                    )
                )
                
                // Form
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .verticalScroll(rememberScrollState())
                        .padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(20.dp)
                ) {
                    // Cover Image URL
                    FormField(
                        label = "Cover Image (URL)",
                        value = coverImageURL,
                        onValueChange = { coverImageURL = it },
                        placeholder = "Cover Image URL"
                    )
                    
                    // Book Title
                    FormField(
                        label = "Book Title",
                        value = bookTitle,
                        onValueChange = {
                            bookTitle = it
                            bookTitleError = null
                        },
                        placeholder = "Book Title",
                        isError = bookTitleError != null,
                        errorMessage = bookTitleError
                    )
                    
                    // Genre
                    FormField(
                        label = "Genre",
                        value = genre,
                        onValueChange = { genre = it },
                        placeholder = "Book genre"
                    )
                    
                    // Author Name
                    FormField(
                        label = "Author Name",
                        value = authorName,
                        onValueChange = {
                            authorName = it
                            authorNameError = null
                        },
                        placeholder = "Book Author",
                        isError = authorNameError != null,
                        errorMessage = authorNameError
                    )
                    
                    // Status
                    Column {
                        Text(
                            "Status",
                            fontSize = 14.sp,
                            color = Color.Black,
                            modifier = Modifier.padding(bottom = 8.dp)
                        )
                        StatusSelector(
                            selectedStatus = readingStatus,
                            onStatusSelected = { readingStatus = it }
                        )
                    }
                    
                    // Start Date
                    FormField(
                        label = "Start date",
                        value = dateFormat.format(startDate),
                        onValueChange = { },
                        placeholder = "Start date of reading",
                        readOnly = true,
                        onClick = {
                            calendar.time = startDate
                            DatePickerDialog(
                                context,
                                { _, year, month, dayOfMonth ->
                                    calendar.set(year, month, dayOfMonth, 0, 0, 0)
                                    calendar.set(Calendar.MILLISECOND, 0)
                                    startDate = calendar.time
                                },
                                calendar.get(Calendar.YEAR),
                                calendar.get(Calendar.MONTH),
                                calendar.get(Calendar.DAY_OF_MONTH)
                            ).show()
                        }
                    )
                    
                    // Rating
                    FormField(
                        label = "Rating (0 - 5 stars)",
                        value = starRating,
                        onValueChange = {
                            starRating = it
                            starRatingError = null
                        },
                        placeholder = "Rating",
                        isError = starRatingError != null,
                        errorMessage = starRatingError
                    )
                    
                    // Rating Description
                    FormField(
                        label = "Rating Description",
                        value = reviewDescription,
                        onValueChange = { reviewDescription = it },
                        placeholder = "Description",
                        minLines = 3
                    )
                    
                    Spacer(modifier = Modifier.height(16.dp))
                    
                    // Update Button
                    Button(
                        onClick = {
                            // Validation
                            var isValid = true
                            bookTitleError = null
                            authorNameError = null
                            starRatingError = null
                            
                            if (bookTitle.isBlank()) {
                                bookTitleError = "Book title is required"
                                isValid = false
                            }
                            
                            if (authorName.isBlank()) {
                                authorNameError = "Author name is required"
                                isValid = false
                            }
                            
                            if (starRating.isNotBlank()) {
                                val rating = starRating.toFloatOrNull()
                                if (rating == null || rating < 0 || rating > 5) {
                                    starRatingError = "Rating must be between 0 and 5"
                                    isValid = false
                                }
                            }
                            
                            if (isValid) {
                                val updatedBook = book.copy(
                                    bookTitle = bookTitle.trim(),
                                    authorName = authorName.trim(),
                                    genre = genre.trim(),
                                    readingStatus = readingStatus,
                                    starRating = starRating.toFloatOrNull() ?: 0f,
                                    reviewDescription = reviewDescription.trim(),
                                    coverImageURL = coverImageURL.trim(),
                                    startDate = startDate
                                )
                                onConfirm(updatedBook)
                            }
                        },
                        modifier = Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(8.dp),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = Color(0xFF2196F3)
                        )
                    ) {
                        Text(
                            "Update Review",
                            fontSize = 16.sp,
                            modifier = Modifier.padding(vertical = 12.dp)
                        )
                    }
                }
            }
        }
    }
}
